pkg/cosign/image_signer_test.go

```diff
@@ -0,0 +1,575 @@
+package cosign
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http/httptest"
+	"net/url"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/BurntSushi/toml"
+	"github.com/buildpacks/lifecycle"
+	"github.com/google/go-containerregistry/pkg/authn"
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/google/go-containerregistry/pkg/registry"
+	"github.com/google/go-containerregistry/pkg/v1/random"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+	"github.com/sclevine/spec"
+	"github.com/sigstore/cosign/cmd/cosign/cli"
+	"github.com/sigstore/cosign/cmd/cosign/cli/download"
+	sigstoreCosign "github.com/sigstore/cosign/pkg/cosign"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestImageSigner(t *testing.T) {
+	spec.Run(t, "Test Cosign Image Signer Main", testImageSigner)
+}
+
+func testImageSigner(t *testing.T, when spec.G, it spec.S) {
+	var (
+		report            lifecycle.ExportReport
+		reader            *os.File
+		writer            *os.File
+		scanner           *bufio.Scanner
+		expectedImageName string
+		stopRegistry      func()
+		imageCleanup      func()
+		repo              string
+		testCtx           context.Context
+		testCtxCancel     context.CancelFunc
+	)
+
+	it.Before(func() {
+		testCtx, testCtxCancel = context.WithTimeout(context.Background(), time.Minute)
+		scanner, reader, writer = mockLogger(t)
+		repo, stopRegistry = reg(t)
+
+		expectedImageName = path.Join(repo, "test-cosign-image")
+
+		imageCleanup = pushRandomImage(t, expectedImageName)
+	})
+
+	it.After(func() {
+		testCtxCancel()
+		stopRegistry()
+		imageCleanup()
+		resetLogger(reader, writer)
+	})
+
+	when("#Sign", func() {
+		var (
+			secretKey1     string
+			publicKey1     string
+			publicKey2     string
+			passwordFile1  string
+			passwordFile2  string
+			secretLocation string
+		)
+
+		when("signing occurs", func() {
+			it.Before(func() {
+				// Override secretLocation for test
+				secretLocation = createCosignKeyFiles(t)
+
+				secretKey1 = path.Join(secretLocation, "secret-name-1", "cosign.key")
+				publicKey1 = path.Join(secretLocation, "secret-name-1", "cosign.pub")
+				publicKey2 = path.Join(secretLocation, "secret-name-2", "cosign.pub")
+				passwordFile1 = path.Join(secretLocation, "secret-name-1", "cosign.password")
+				passwordFile2 = path.Join(secretLocation, "secret-name-2", "cosign.password")
+
+				report = createReportToml(t, expectedImageName)
+
+				os.Unsetenv(cosignRepositoryEnv)
+				os.Unsetenv(cosignDockerMediaTypesEnv)
+			})
+
+			it("signs images", func() {
+				cliSignCmdCallCount := 0
+				password1Count := 0
+				password2Count := 0
+
+				cliSignCmd := func(ctx context.Context, ko cli.KeyOpts, annotations map[string]interface{}, imageRef, certPath string, upload bool, payloadPath string, force, recursive bool) error {
+					t.Helper()
+					assert.Equal(t, testCtx, ctx)
```

Comment:
I really like that these tests actually validate the signature. I think the additional validation of the values passed to the `cliSignCmd` here might be unnecessary (and may make it hard to use new api versions).

I think it is fine for now but, perhaps something to think about in the future.